duktape/src/main/jni/JavaMethod.cpp

```diff
@@ -19,119 +19,160 @@
+JavaMethod::MethodBody getMethodBody(JNIEnv* env, jmethodID methodId, jobject returnType);
 
 } // anonymous namespace
 
-JavaMethod::JavaMethod(JNIEnv* env, jobject method)
-    : m_methodId(env->FromReflectedMethod(method)) {
+JavaMethod::JavaMethod(JNIEnv* env, jobject method) {
   jclass methodClass = env->GetObjectClass(method);
 
-  const jmethodID getName = env->GetMethodID(methodClass, "getName", "()Ljava/lang/String;");
-  const JString methodName(env, static_cast<jstring>(env->CallObjectMethod(method, getName)));
-  m_name = methodName;
-
-  const jmethodID getReturnType =
-      env->GetMethodID(methodClass, "getReturnType", "()Ljava/lang/Class;");
-  jobject returnType = env->CallObjectMethod(method, getReturnType);
-  m_returnType = getType(env, returnType);
-
   const jmethodID getParameterTypes =
       env->GetMethodID(methodClass, "getParameterTypes", "()[Ljava/lang/Class;");
   jobjectArray parameterTypes =
       static_cast<jobjectArray>(env->CallObjectMethod(method, getParameterTypes));
   const jsize numArgs = env->GetArrayLength(parameterTypes);
-  m_parameterTypes.reserve(numArgs);
+  m_argumentLoaders.reserve(numArgs);
   for (jsize i = 0; i < numArgs; ++i) {
-    m_parameterTypes.push_back(getType(env, env->GetObjectArrayElement(parameterTypes, i)));
+    m_argumentLoaders.push_back(
+        getArgumentLoader(env, env->GetObjectArrayElement(parameterTypes, i)));
   }
+
+  const jmethodID getReturnType =
+      env->GetMethodID(methodClass, "getReturnType", "()Ljava/lang/Class;");
+  jobject returnType = env->CallObjectMethod(method, getReturnType);
+
+  jmethodID methodId = env->FromReflectedMethod(method);
+  m_methodBody = getMethodBody(env, methodId, returnType);
 }
 
-duk_ret_t JavaMethod::invoke(duk_context *ctx, JNIEnv *env, jobject javaThis) const {
-  if (duk_get_top(ctx) != m_parameterTypes.size()) {
+duk_ret_t JavaMethod::invoke(duk_context* ctx, JNIEnv* env, jobject javaThis) const {
+  if (duk_get_top(ctx) != m_argumentLoaders.size()) {
     duk_error(ctx, DUK_ERR_API_ERROR, "wrong number of arguments");
   }
-  std::vector<jvalue> args(m_parameterTypes.size());
-  std::transform(m_parameterTypes.begin(), m_parameterTypes.end(), args.rbegin(),
-                 [env, ctx](Type t) { return popJavaType(env, ctx, t); });
-
-  jvalue returnValue;
-  switch (m_returnType) {
-    case Void:
-      returnValue.l = nullptr;
-      env->CallVoidMethodA(javaThis, m_methodId, args.data());
-      break;
-
-    case String:
-      returnValue.l = env->CallObjectMethodA(javaThis, m_methodId, args.data());
-      break;
-
-    default:
-      return DUK_RET_INTERNAL_ERROR;
+
+  std::vector<jvalue> args(m_argumentLoaders.size());
+  // Load the arguments off the stack and convert to java types.
+  // Note we're going backwards since the last argument is at the top of the stack.
+  std::transform(m_argumentLoaders.rbegin(), m_argumentLoaders.rend(), args.rbegin(),
```

Comment:
Yeah, the trick is that we're iterating over the vector backwards, since we're popping the last java argument first, and using `rbegin` in a transform call seemed more straightforward at the time than an indexed loop counting down.  I agree, though, it's a little cumbersome.  I can make it a for loop with a numeric index.